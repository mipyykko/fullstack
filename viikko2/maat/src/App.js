import React from 'react'
import axios from 'axios'

const Countries = ({ countries, setFilter }) => {
    if (countries.length > 10) {
        return (
            <div>Too many matches</div>
        )
    } else if (countries.length > 1) {
        return (
            <div>
            <ul>
                {countries.map(country => <li key={country.name}><div onClick={() => setFilter(country.name)}>{country.name}</div></li>)}
            </ul>    
            </div>
        )
    } else if (countries.length === 1) {
        const country = countries[0]
        return (
            <div>
                <h2>{country.nativeName} {country.name}</h2>
                <img src={country.flag} alt={country.name} width='80%' />
                <div>capital: {country.capital}</div>
                <div>population: {country.population}</div>
            </div>
        )
    }

    return(
        <div>No matches!</div>
    )
}

export default class App extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            filter: '',
            countries: [],
            filteredCountries: []
        }
    }

    updateFilter = (filter) => {
        this.setState({ 
            filteredCountries: 
                this.state.countries.length > 0 
              ? this.state.countries.filter(
                  country => country.name.trim().toLowerCase()
                                .indexOf(filter.trim().toLowerCase()) >= 0)
              : [],
            filter
         })
    }

    handleCountryChange = (event) => {
        this.updateFilter(event.target.value)
    }

    componentWillMount() {
        axios.get('https://restcountries.eu/rest/v2/all')
            .then(response => {
                this.setState({ countries: response.data })
            })
            .catch(error => console.err(error))
    }

    render() {
        return(
            <div>
                <div>
                    find countries: 
                    <input value={this.state.filter}
                        onChange={this.handleCountryChange}
                    />
                </div>
                <Countries countries={this.state.filteredCountries} setFilter={this.updateFilter} /> 
            </div>
        )
    }

}