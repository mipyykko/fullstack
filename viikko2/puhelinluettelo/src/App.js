import React from 'react';

class App extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      persons: [
        { name: 'Arto Hellas' }
      ],
      newName: ''
    }
  }

  handleNameChange = (event) => {
      this.setState({ newName: event.target.value })
  }

  addNumber = (event) => {
    event.preventDefault()
    var exists = Object.keys(this.state.persons)
        .some(key => this.state.persons[key].name === this.state.newName)
    if (exists) {
        this.setState({ newName: '' })
    } else {
        this.setState({ persons: [...this.state.persons, { name: this.state.newName }], newName: ''})
    }
  }

  render() {
    return (
      <div>
        <h2>Puhelinluettelo</h2>
        <form onSubmit={this.addNumber}>
          <div>
            nimi: <input 
                value={this.state.newName} 
                onChange={this.handleNameChange} />
          </div>
          <div>
            <button type="submit">lisää</button>
          </div>
        </form>
        <h2>Numerot</h2>
        <ul>
            {this.state.persons.map((person) => <li key={person.name}>{person.name}</li>)}
        </ul>
      </div>
    )
  }
}

export default App