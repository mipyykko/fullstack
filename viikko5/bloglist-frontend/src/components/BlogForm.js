import React from 'react'
import Blog from './Blog'

const BlogForm = ({ blogs, handleLike, handleDelete, user }) => (
  <div>
    <h2>Blogs</h2>
    {blogs.map(blog =>
      <Blog
        key={blog.id}
        blog={blog}
        handleLike={() => handleLike(blog.id)}
        handleDelete={() => handleDelete(blog.id)}
        deletable={
          blog.user === undefined ||
          (blog && user && blog.user.username === user.username)
        }
      />
    )}
  </div>
)

export default BlogForm
