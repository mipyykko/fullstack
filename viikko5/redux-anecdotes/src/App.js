import React from 'react';


class App extends React.Component {
  store = this.props.store

  constructor(props) {
    super(props)
    this.state = { anecdote: '' }
  }

  upvote = (id) => {
    this.store.dispatch({
      type: 'UPVOTE',
      data: { id }
    })
  }

  handleAnecdoteFormChange = (event) => {
    this.setState({ anecdote: event.target.value })
  }

  handleAnecdoteCreate = (event) => {
    event.preventDefault()
    this.store.dispatch({
      type: 'NEW_ANECDOTE',
      data: { anecdote: this.state.anecdote }
    })
    this.setState({Â anecdote: '' })
  }

  render() {
    const anecdotes = this.store.getState()
    anecdotes.sort((a, b) => b.votes - a.votes)

    return (
      <div>
        <h2>Anecdotes</h2>
        {anecdotes.map(anecdote=>
          <div key={anecdote.id}>
            <div>
              {anecdote.content} 
            </div>
            <div>
              has {anecdote.votes}
              <button onClick={() => this.upvote(anecdote.id)}>vote</button>
            </div>
          </div>
        )}
        <h2>create new</h2>
        <form onSubmit={this.handleAnecdoteCreate}>
          <div><input 
            type="text"
            name="anecdote"
            value={this.state.anecdote}
            onChange={this.handleAnecdoteFormChange}
          /></div>
          <button type="submit">create</button> 
        </form>
      </div>
    )
  }
}

export default App