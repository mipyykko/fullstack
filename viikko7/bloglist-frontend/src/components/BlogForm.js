import React from 'react'
import { connect } from 'react-redux'
import { likeBlog, deleteBlog } from '../reducers/blogReducer'
import { Table } from 'semantic-ui-react'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'

class BlogForm extends React.Component {

  render() {
    const blogs = this.props.blogs

    return(
      <div>
        <h2>Blogs</h2>
        <Table striped celled>
          <Table.Body>
            {blogs.map(blog =>
              <Table.Row key={blog.id}>
                <Table.Cell>
                  <Link to={`/blogs/${blog.id}`}>{blog.title} by {blog.author}</Link>
                </Table.Cell>
              </Table.Row>
            )}
          </Table.Body>
        </Table>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return({
    blogs: state.blogs.sort((a, b) => b.likes - a.likes)
  })
}

const mapDispatchToProps = {
  likeBlog,
  deleteBlog
}

BlogForm.propTypes = {
  likeBlog: PropTypes.func.isRequired,
  deleteBlog: PropTypes.func.isRequired,
  blogs: PropTypes.array.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(BlogForm)
