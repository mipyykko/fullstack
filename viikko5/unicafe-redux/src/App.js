import React from 'react';
import './App.css';

let store

const Statistiikka = () => {
  const data = store.getState()

  const palautteita = Object.keys(data).reduce((prev, key) => prev + data[key], 0)
  const pos = data.good / palautteita * 100
  const keskiarvo = (data.good - data.bad) / palautteita

  if (palautteita === 0) {
    return (
      <div>
        <h2>statistiikka</h2>
        <div>ei yhtään palautetta annettu</div>
      </div>
    )
  }

  return (
    <div>
      <h2>statistiikka</h2>
      <table>
        <tbody>
          <tr>
            <td>hyvä</td>
            <td>{data.good}</td>
          </tr>
          <tr>
            <td>neutraali</td>
            <td>{data.ok}</td>
          </tr>
          <tr>
            <td>huono</td>
            <td>{data.bad}</td>
          </tr>
          <tr>
            <td>keskiarvo</td>
            <td>{keskiarvo.toFixed(2)}</td>
          </tr>
          <tr>
            <td>positiivisia</td>
            <td>{pos.toFixed(2)}%</td>
          </tr>
        </tbody>
      </table>

      <button onClick={e => store.dispatch({ type: 'ZERO' })}>nollaa tilasto</button>
    </div >
  )
}

class App extends React.Component {
  constructor(props) {
    super(props)
    store = props.store
  }

  klik = (nappi) => () => {
    store.dispatch({ type: nappi })
  }

  render() {
    return (
      <div>
        <h2>anna palautetta</h2>
        <button onClick={this.klik('GOOD')}>hyvä</button>
        <button onClick={this.klik('OK')}>neutraali</button>
        <button onClick={this.klik('BAD')}>huono</button>
        <Statistiikka />
      </div>
    )
  }
}

export default App;
