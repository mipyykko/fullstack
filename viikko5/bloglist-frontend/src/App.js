import React from 'react'
import LoginForm from './components/LoginForm'
import BlogForm from './components/BlogForm'
import CreateBlogForm from './components/CreateBlogForm'
import Notification from './components/Notification'
import Togglable from './components/Togglable'
import blogService from './services/blogs'
import loginService from './services/login'

class App extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      blogs: [],
      user: null
    }
  }

  componentDidMount() {
    blogService.getAll().then(blogs =>
    {
      this.setState({ blogs })
    }
    )

    const loggedUserJSON = window.localStorage.getItem('loggedInUser')
    if (loggedUserJSON) {
      const user = JSON.parse(loggedUserJSON)
      this.setState({ user })
      blogService.setToken(user.token)
    }
  }

  login = async (event) => {
    event.preventDefault()
    try {
      const user = await loginService.login({
        username: this.state.username,
        password: this.state.password
      })

      window.localStorage.setItem('loggedInUser', JSON.stringify(user))
      blogService.setToken(user.token)
      this.setState({
        notification: `${user.name} logged in`,
        username: '', password: '', user
      })
      setTimeout(() => {
        this.setState({ notification: null })
      }, 2000)
    } catch (exception) {
      this.setState({ error: 'wrong user or password' })
      setTimeout(() => {
        this.setState({ error: null })
      }, 2000)
    }
  }

  logout = () => {
    window.localStorage.removeItem('loggedInUser')
    this.setState({
      notification: `${this.state.user.name} logged out`,
      user: null
    })
    setTimeout(() => {
      this.setState({ notification: null })
    }, 2000)
  }

  handleLoginFormChange = (event) => {
    this.setState({ [event.target.name]: event.target.value })
  }

  createBlog = async (event) => {
    event.preventDefault()
    try {
      const newBlog = await blogService.create({
        title: this.state.title,
        author: this.state.author,
        url: this.state.url
      })
      if (newBlog !== null) {
        this.setState({ blogs: [...this.state.blogs, newBlog ] })
      }
      this.setState({
        notification: `blog ${newBlog.title} created`,
        title: '', author: '', url: ''
      })
      this.createBlogForm.toggleVisibility()
      setTimeout(() => {
        this.setState({ notification: null })
      }, 2000)
    } catch (exception) {
      this.setState({ error: 'missing fields' })
      setTimeout(() => {
        this.setState({ error: null })
      }, 2000)
    }
  }

  handleCreateBlogFormChange = (event) => {
    this.setState({ [event.target.name]: event.target.value })
  }

  handleLike = async (id) => {
    try {
      const likedBlog = await blogService.like(id)

      if (likedBlog === null) {
        throw new Error('wrong id')
      }

      const likedIdx = this.state.blogs.findIndex(blog => id === blog.id)
      this.setState({
        blogs: this.state.blogs.slice(0, likedIdx)
          .concat([{ ...this.state.blogs[likedIdx],
            likes: likedBlog.likes }])
          .concat(this.state.blogs.slice(likedIdx + 1))
      })
    } catch (exception) {
      this.setState({ error: 'wrong id?' })
      setTimeout(() => {
        this.setState({ error: null })
      }, 2000)

      console.log(exception)
    }
  }

  toggleExpanded = (ref) => {
    console.log(ref)
    let expanded = this.state.expanded
    expanded[ref] = !expanded[ref]
    this.setState({ expanded })
  }

  render() {
    const loggedIn = () => (
      <div>
        {this.state.user.name} logged in <button onClick={this.logout}>Logout</button>
      </div>
    )

    return (
      <div>
        <h2>blogs</h2>
        <Notification
          notification={this.state.notification}
          error={this.state.error}
        />
        {this.state.user === null ?
          <Togglable buttonLabel="log in">
            <LoginForm
              handleLogin={this.login}
              handleLoginFormChange={this.handleLoginFormChange}
              username={this.state.username}
              password={this.state.password}
            />
          </Togglable> :
          <div>
            {loggedIn()}
            <Togglable
              buttonLabel="new blog"
              ref={component => this.createBlogForm = component}>
              <CreateBlogForm
                handleCreateBlog={this.createBlog}
                handleCreateBlogFormChange={this.handleCreateBlogFormChange}
                title={this.state.title}
                author={this.state.author}
                url={this.state.url}
              />
            </Togglable>
            <BlogForm
              blogs={this.state.blogs}
              handleLike={this.handleLike}
            />
          </div>
        }
      </div>
    )
  }
}

export default App
