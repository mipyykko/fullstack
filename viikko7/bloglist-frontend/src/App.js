import React from 'react'
import { BrowserRouter as Router, Route, Link } from 'react-router-dom'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import { Menu, Button } from 'semantic-ui-react'

import LoginForm from './components/LoginForm'
import BlogForm from './components/BlogForm'
import CreateBlogForm from './components/CreateBlogForm'
import Notifications from './components/Notifications'
import Users from './components/Users'
import User from './components/User'
import Blog from './components/Blog'

import { initBlogs } from './reducers/blogReducer'
import { initUsers } from './reducers/userReducer'
import { setUser, logoutUser } from './reducers/loginReducer'

class App extends React.Component {

  componentDidMount() {
    this.props.initBlogs()
    this.props.initUsers()

    const loggedUserJSON = window.localStorage.getItem('loggedInUser')

    if (loggedUserJSON) {
      const user = JSON.parse(loggedUserJSON)
      this.props.setUser(user)
    }
  }


  logout = () => {
    this.props.logoutUser()
  }

  render() {
    const loggedIn = () => (
      <div>
        <em>{this.props.login.name}</em> logged in <Button onClick={this.logout}>Logout</Button>
      </div>
    )

    return (
      <div>
        <Router>
          <div>
            <h2>blogs</h2>
            <Menu inverted>
              <Menu.Item link>
                <Link to="/">blogs</Link>
              </Menu.Item>
              <Menu.Item link>
                <Link to="/users">users</Link>
              </Menu.Item>
              <Menu.Item link>
                {this.props.login.name &&
                  loggedIn()}
              </Menu.Item>
            </Menu>
            <Notifications />
            {!this.props.login.token ?
              <LoginForm /> :
              <div>
                <CreateBlogForm />

                <Route exact path="/" render={() => {
                  return(
                    <div className="blogContent">
                      <BlogForm />
                    </div>
                  )}} />
                <Route exact path="/users" render={() =>
                  <Users />}
                />
                <Route exact path="/users/:id" render={({ match }) =>
                  <User id={match.params.id} />
                }/>
                <Route exact path="/blogs/:id" render={({ match }) =>
                  <Blog id={match.params.id} />
                }/>
              </div>
            }
          </div>
        </Router>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return ({
    login: state.login
  })
}

const mapDispatchToProps = {
  initBlogs,
  initUsers,
  setUser,
  logoutUser
}

App.propTypes = {
  initBlogs: PropTypes.func.isRequired,
  initUsers: PropTypes.func.isRequired,
  setUser: PropTypes.func.isRequired,
  logoutUser: PropTypes.func.isRequired,
  login: PropTypes.object.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(App)
