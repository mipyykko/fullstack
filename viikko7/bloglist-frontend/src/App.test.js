import React from 'react'
import { mount } from 'enzyme'
import App from './App'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import BlogForm from './components/BlogForm'
import renderer from 'react-test-renderer'
jest.mock('./services/blogs') // eslint-disable-line
jest.mock('./services/users') // eslint-disable-line
import blogService from './services/blogs'
// import userService from './services/users'
import configureMockStore from 'redux-mock-store'

const puppeteer = require('puppeteer')

const middlewares = [ thunk ]
const mockStore = configureMockStore(middlewares)

describe('<App />', () => {
  let app, store

  const initialState = {
    blogs: blogService.blogs,
    users: [],
    login: {},
    notifications: []
  }

  describe('user is not logged in', () => {
    beforeEach(() => {
      store = mockStore(initialState)
      app = mount(<Provider store={store}><App /></Provider>)
    })

    it('does not render blogs without login', () => {
      app.update()
      const blogComponents = app.find(BlogForm)
      expect(blogComponents.length).toEqual(0)
    })

    it('renders correctly', () => {
      const tree = renderer
        .create(<Provider store={store}><App /></Provider>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('user is logged in', async () => {
    let store

    beforeEach(async () => {
      const user = {
        username: 'asdf',
        name: 'hdfg',
        token: '66666'
      }
      store = await mockStore({ ...initialState, login: user })

      window.localStorage.setItem('loggedInUser', JSON.stringify(user))

      app = mount(<Provider store={store}><App /></Provider>)
    })

    it('renders blogs after login', () => {
      app.update()
      const blogComponents = app.find(BlogForm)
      expect(blogComponents.length).toEqual(blogService.blogs.length)
    })

    it('renders correctly', () => {
      const tree = renderer
        .create(<Provider store={store}><App /></Provider>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

  })

  describe('puppeteer tests', () => {
    let page, browser

    beforeEach(async () => {
      browser = await puppeteer.launch({
        headless: true
      })
      page = await browser.newPage()
      await page.goto('http://localhost:3000')
    })

    it('renders with browser and has a menu', async () => {
      const bodyContent = await page.$eval('body', el => el.textContent)
      const menu = await page.$eval('.ui', el => (el ? true : false))

      expect(bodyContent.includes('Blog application')).toBe(true)
      expect(menu).toBe(true)
    }, 10000)

    afterAll(() => {
      browser.close()
    })
  })
})