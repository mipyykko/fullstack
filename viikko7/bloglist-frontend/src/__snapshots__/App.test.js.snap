// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> user is logged in renders correctly 1`] = `
<div>
  <div>
    <h2>
      Blog application
    </h2>
    <div
      className="ui inverted menu"
    >
      <div
        className="link item"
        onClick={[Function]}
      >
        <a
          href="/"
          onClick={[Function]}
        >
          blogs
        </a>
      </div>
      <div
        className="link item"
        onClick={[Function]}
      >
        <a
          href="/users"
          onClick={[Function]}
        >
          users
        </a>
      </div>
      <div
        className="link item"
        onClick={[Function]}
      >
        <div>
          <em>
            hdfg
          </em>
           logged in 
          <button
            className="ui button"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            tabIndex={undefined}
          >
            Logout
          </button>
        </div>
      </div>
    </div>
    <div />
    <div>
      <div>
        <div
          style={
            Object {
              "display": "",
            }
          }
        >
          <button
            className="ui button"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            tabIndex={undefined}
          >
            new blog
          </button>
        </div>
        <div
          style={
            Object {
              "display": "none",
            }
          }
        >
          <div>
            <h2>
              Create new blog
            </h2>
            <form
              action={undefined}
              className="ui form"
              onSubmit={[Function]}
            >
              <div
                className="field"
              >
                <div
                  className="field"
                >
                  <label
                    htmlFor={undefined}
                  >
                    title
                  </label>
                  <div
                    className="ui input"
                    content={undefined}
                  >
                    <input
                      disabled={undefined}
                      name="title"
                      onChange={[Function]}
                      placeholder="title"
                      required={undefined}
                      tabIndex={undefined}
                      type="text"
                    />
                  </div>
                </div>
              </div>
              <div
                className="field"
              >
                <div
                  className="field"
                >
                  <label
                    htmlFor={undefined}
                  >
                    author
                  </label>
                  <div
                    className="ui input"
                    content={undefined}
                  >
                    <input
                      disabled={undefined}
                      name="author"
                      onChange={[Function]}
                      placeholder="author"
                      required={undefined}
                      tabIndex={undefined}
                      type="text"
                    />
                  </div>
                </div>
              </div>
              <div
                className="field"
              >
                <div
                  className="field"
                >
                  <label
                    htmlFor={undefined}
                  >
                    url
                  </label>
                  <div
                    className="ui input"
                    content={undefined}
                  >
                    <input
                      disabled={undefined}
                      name="url"
                      onChange={[Function]}
                      placeholder="url"
                      required={undefined}
                      tabIndex={undefined}
                      type="text"
                    />
                  </div>
                </div>
              </div>
              <button
                className="ui button"
                disabled={undefined}
                onClick={[Function]}
                role="button"
                tabIndex={undefined}
                type="submit"
              >
                Create
              </button>
            </form>
          </div>
          <button
            className="ui button"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            tabIndex={undefined}
          >
            cancel
          </button>
        </div>
      </div>
      <div
        className="blogContent"
      >
        <div>
          <h2>
            Blogs
          </h2>
          <table
            className="ui celled striped table"
          >
            <tbody
              className=""
            >
              <tr
                className=""
              >
                <td
                  className=""
                >
                  <a
                    href="/blogs/abcdef56"
                    onClick={[Function]}
                  >
                    asdf fdfdf
                     by 
                    hhhhh6
                  </a>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<App /> user is not logged in renders correctly 1`] = `
<div>
  <div>
    <h2>
      Blog application
    </h2>
    <div
      className="ui inverted menu"
    >
      <div
        className="link item"
        onClick={[Function]}
      >
        <a
          href="/"
          onClick={[Function]}
        >
          blogs
        </a>
      </div>
      <div
        className="link item"
        onClick={[Function]}
      >
        <a
          href="/users"
          onClick={[Function]}
        >
          users
        </a>
      </div>
      <div
        className="link item"
        onClick={[Function]}
      >
        
      </div>
    </div>
    <div />
    <div>
      <div
        style={
          Object {
            "display": "",
          }
        }
      >
        <button
          className="ui button"
          disabled={undefined}
          onClick={[Function]}
          role="button"
          tabIndex={undefined}
        >
          log in
        </button>
      </div>
      <div
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div>
          <h2>
            Log in
          </h2>
          <form
            action={undefined}
            className="ui form"
            onSubmit={[Function]}
          >
            <div
              className="field"
            >
              <div
                className="field"
              >
                <label
                  htmlFor={undefined}
                >
                  username
                </label>
                <div
                  className="ui input"
                  content={undefined}
                >
                  <input
                    disabled={undefined}
                    name="username"
                    onChange={[Function]}
                    placeholder="username"
                    required={undefined}
                    tabIndex={undefined}
                    type="text"
                  />
                </div>
              </div>
            </div>
            <div
              className="field"
            >
              <div
                className="field"
              >
                <label
                  htmlFor={undefined}
                >
                  password
                </label>
                <div
                  className="ui input"
                  content={undefined}
                >
                  <input
                    disabled={undefined}
                    name="password"
                    onChange={[Function]}
                    placeholder="password"
                    required={undefined}
                    tabIndex={undefined}
                    type="password"
                  />
                </div>
              </div>
            </div>
            <button
              className="ui button"
              disabled={undefined}
              onClick={[Function]}
              role="button"
              tabIndex={undefined}
              type="submit"
            >
              login
            </button>
          </form>
        </div>
        <button
          className="ui button"
          disabled={undefined}
          onClick={[Function]}
          role="button"
          tabIndex={undefined}
        >
          cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
